From 63de52a7c717fba30a011cf97b5f5dc215878815 Mon Sep 17 00:00:00 2001
From: Andrea Canciani <ranma42@gmail.com>
Date: Wed, 21 Nov 2012 23:56:52 +0100
Subject: [PATCH] HACK: Disallow block-based textures with sub-block size

League of Legends crashes when using textures with mipmaps smaller
than a block.

Work around this by forcing the failure of the creation of such
textures.

WARNING: THIS IS JUST A HACK, NOT A PROPER FIX

In order to get a proper fix, we should investigate and find out why
(and where, i.e. in wine or in the LoL client) we're overwriting the
heap-allocated data structures (textures, freelist).
---
 dlls/wined3d/utils.c |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/wined3d/utils.c b/dlls/wined3d/utils.c
index 2ade261..7c0a9a4 100644
--- a/dlls/wined3d/utils.c
+++ b/dlls/wined3d/utils.c
@@ -1699,6 +1699,13 @@ UINT wined3d_format_calculate_size(const struct wined3d_format *format, UINT ali
         UINT row_block_count = (width + format->block_width - 1) / format->block_width;
         UINT row_count = (height + format->block_height - 1) / format->block_height;
         size = row_count * (((row_block_count * format->block_byte_count) + alignment - 1) & ~(alignment - 1));
+
+	if (width < format->block_width && height < format->block_height) {
+		WARN("Invalid size: %u x %u -> %u (min %u x %u)\n",
+		     width, height, size, format->block_width, format->block_height);
+
+		size = 0;
+	}
     }
     else
     {
-- 
1.7.10.2 (Apple Git-33)

